include/tuvx/linalg/linalg.inl

```diff
@@ -0,0 +1,85 @@
+
+namespace tuvx {
+
+template <typename T>
+inline std::vector<T> Dot(TridiagonalMatrix<T> A, std::vector<T> x) {
+  std::size_t size = x.size();
+  std::vector<T> v(size);
+  v[0] = A.main_diagonal_[0] * x[0] + A.upper_diagonal_[0] * x[1];
+
+  std::size_t i = 0;
+  for (i = 1; i < size - 1; i++) {
+    v[i] = A.main_diagonal_[i] * x[i] + A.upper_diagonal_[i] * x[i + 1] +
+           A.lower_diagonal_[i - 1] * x[i - 1];
+  }
+  v[i] = A.main_diagonal_[i] * x[i] + A.lower_diagonal_[i - 1] * x[i - 1];
+  return v;
+}
+
+template <typename T>
+inline std::vector<T> TridiagonalSolve(TridiagonalMatrix<T> A,
+                                       std::vector<T> b) {
```

Comment:
Could you make the parameters that doesn't need a copy to `const ref` ? This request applies to all the functions in this PR. 
```suggestion
template <typename T>
inline std::vector<T> TridiagonalSolve(const TridiagonalMatrix<T>& A,
                                       const std::vector<T>& b) {
```

Reply:
The `TridiagonalSolve` code modifies the arguments instead of defining new temporary variables in the function. The rest of the functions in this PR do not do this so I will apply your suggestions to them.